### PR TITLE
Handle aperture photometry when images linked by WCS for advanced cases

### LINKED ISSUE #2139: Ensure aperture is correct when images linked by WCS
**Reporter:** pllim

Consider the following use case: Load two images in Imviz and link them by WCS. The second image has different resolution (e.g., covering the same area in sky but with twice as many pixels) or has different sky orientation.

Do the Subsets (circle, ellipse, rectangle) give correct answers when applied to the second (non-reference) image? If not, how do we fix this (a rectangle might not even be a rectangle anymore when projected on the second image)? If there is no easy fix, should we disable aperture photometry for non-reference images?

To test the seriousness of this situation, measure the correct answers first by loading only the second image in an Imviz session (maybe start with Circle, the easiest of the shapes) and run some aperture photometry. Then draw what you think is the same circle but now with this image as a non-reference image (the reference image should have different sky resolution to test this). Rerun the aperture photometry. Do the aperture sums agree within uncertainty?

The JWST images in the Imviz example notebook might be a good start, as they have different sky resolutions, I think.

[🐱](https://jira.stsci.edu/browse/JDAT-3272)

### Blocked by

* #2190

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### LINKED ISSUE #2265: Recenter tool only looks at reference layer
**Reporter: Camilla Pacifici**

1. I have 2 images loaded with different resolutions (e.g., same object in at pixel 300,500 in image 1 and at pixel 30,50 in image 2).
2. I select a subset. Subset tool says the subset is at 300,500.
3. If I select image 1 and I use the recenter tool, it works fine.
4. If I select image 2 and I use the recenter tool, it tries to recenter on pixel 300,500 instead of at 30,50 and starts wondering around. The user has no ides what is happening.

Need to figure out a way to deal with this.

As a user, in an ideal world, I would expect the subset to automatically shift to the correct position in image 2, but this might get cumbersome if the user keeps switching from one to the other.
Working in sky coordinates might solve the problem, but the pixel position would still need to be obvious to the user.

Any mitigation plans in the meantime are welcome.

[🐱](https://jira.stsci.edu/browse/JDAT-3544)

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### LINKED ISSUE #2357: Translations between astropy sky region angle and photutils sky aperture angle are wrong
**Reporter:** pllim

Despite similar descriptions in docstrings, when I dig into the actual code, one is actually defined against longitude and one latitude, so they are actually off by 90 degrees. Noticed this when I translated one to the other and got different aperture sum values (https://github.com/spacetelescope/jdaviz/issues/2139).

`RectangleSkyRegion` (regions): *The rotation angle of the rectangle, measured anti-clockwise. If set to zero (the default), the width axis is lined up with the longitude axis of the celestial coordinates.*

https://github.com/astropy/regions/blob/501d12d5f5879c49413e20bed90a2d3eb725d5b7/regions/shapes/rectangle.py#L117

```python
        # region sky angles are defined relative to the WCS longitude axis;
        # photutils aperture sky angles are defined as the PA of the
        # semimajor axis (i.e., relative to the WCS latitude axis)
        angle = self.angle - (north_angle - 90 * u.deg)
```

`SkyRectangularAperture` (photutils): *The position angle (in angular units) of the rectangle “width” side. For a right-handed world coordinate system, the position angle increases counterclockwise from North (PA=0).*

https://github.com/astropy/photutils/blob/fc3583639732c1fecd4deaf074d9d873c9904519/photutils/aperture/core.py#L746

```python
            if param == 'theta':
                # photutils aperture sky angles are defined as the PA of
                # the semimajor axis (i.e., relative to the WCS latitude
                # axis). region sky angles are defined relative to the WCS
                # longitude axis.
                value = (value * u.rad) - angle.to(u.rad)
```

We should do due diligence and double check the theta code for pixel regions too.

`RectanglePixelRegion` (regions): *The rotation angle of the rectangle, measured anti-clockwise. If set to zero (the default), the width axis is lined up with the x axis.*

https://github.com/astropy/regions/blob/501d12d5f5879c49413e20bed90a2d3eb725d5b7/regions/shapes/rectangle.py#L126C11-L126C11

```python
        cos_angle = np.cos(self.angle)  # self.angle is a Quantity
        sin_angle = np.sin(self.angle)  # self.angle is a Quantity
        dx1 = abs(w2 * cos_angle - h2 * sin_angle)
        dy1 = abs(w2 * sin_angle + h2 * cos_angle)
        dx2 = abs(w2 * cos_angle + h2 * sin_angle)
        dy2 = abs(w2 * sin_angle - h2 * cos_angle)
```

`RectangularAperture` (photutils): *The rotation angle as an angular quantity (Quantity or Angle) or value in radians (as a float) from the positive x axis. The rotation angle increases counterclockwise.*

https://github.com/astropy/photutils/blob/fc3583639732c1fecd4deaf074d9d873c9904519/photutils/aperture/rectangle.py#L118-L123

```python
        sin_theta = math.sin(theta)
        cos_theta = math.cos(theta)
        x_extent1 = abs((half_width * cos_theta) - (half_height * sin_theta))
        x_extent2 = abs((half_width * cos_theta) + (half_height * sin_theta))
        y_extent1 = abs((half_width * sin_theta) + (half_height * cos_theta))
        y_extent2 = abs((half_width * sin_theta) - (half_height * cos_theta))
```

[🐱](https://jira.stsci.edu/browse/JDAT-3698)

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to investigate and hopefully fix the bug of aperture photometry not able to provide correct result in certain cases, e.g.:

* When images have different pixel scales.
* When Subset no longer retains original shape after being projected onto a non-reference image.

* Fixes #2139 
* Fixes #2265 
* Fixes #2357

### Blocked by

* https://github.com/spacetelescope/jdaviz/pull/2164

### TODO

- [x] Remove warning added in https://github.com/spacetelescope/jdaviz/pull/2167 about the aperture (distortion warning should stay).
- [x] Implement the fix.
  - Can we handle this by passing around sky regions in the background?
  - ~Can we also fix Edit Subset plugin by translating sky region to pixel of the selected image? Would need new dropdown for the plugin.~
- [x] See if we can also address https://github.com/spacetelescope/jdaviz/issues/2265
  - [x] Add tests

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? [🐱](https://jira.stsci.edu/browse/JDAT-3272) [🐱](https://jira.stsci.edu/browse/JDAT-3544) [🐱](https://jira.stsci.edu/browse/JDAT-3698)

### CODE REVIEW

File: notebooks/concepts/imviz_advanced_aper_phot.ipynb
Comment (pllim) [inline notebooks/concepts/imviz_advanced_aper_phot.ipynb:352]: This would give:

Screenshot 2023-04-14 173003

This corresponds to the following images loaded into Imviz and linked by WCS:

Screenshot 2023-04-14 173442

Comment (pllim) [reply notebooks/concepts/imviz_advanced_aper_phot.ipynb:352]: I don't know how to:

* pass in arbitrary mask shape (and maybe its bounding box if there is a way to extract mask without reading in the whole image) into `photutils.aperture.ApertureStats` -- @larrybradley ?
* extract subset mask in glue without reading in all the data into memory -- @astrofrog ?

**Comment by pllim** [comment]
@larrybradley and I discussed several options. The most scientifically correct way is still using sky regions (as suggested by @bmorris3). However, this does disable the plugin for any data without WCS.

Hence, Larry also requested that we put in extra logic to still fall back to the old way of doing things (pixel regions) if image does not have WCS and is the reference image for the Subset. But is that the same as a viewer reference image (have to check what becomes the viewer reference image when you only display a non-reference image in the viewer)?

While we're at it, I guess technically we can always falls back to using pixel regions if we detect the link is in pixel space whether the data of interest has WCS or not. Keep in mind that even when Imviz is asked to link by WCS and the reference data has WCS, any image without WCS still links to the reference data by pixel.

And since glue-astronomy does not handle sky regions (see https://github.com/glue-viz/glue-astronomy/issues/89), we would have to rely on our own [region_translators.py](https://github.com/spacetelescope/jdaviz/blob/main/jdaviz/core/region_translators.py).

File: notebooks/concepts/imviz_advanced_aper_phot.ipynb
Comment (pllim) [reply notebooks/concepts/imviz_advanced_aper_phot.ipynb:352]: @larrybradley , to answer your question, if I draw a rectangle on the third image (down-sampled and rotated), the rectangle comes back warped as if I drew it on the reference image, which is what I suspected would happen.

Screenshot 2023-04-17 172540

**Comment by pllim** [comment]
For completeness, Larry said photutils would ignore distortion anyway (that is, if you pass in a circular aperture and WCS with distortion, it would still be circular when extracting the data array, not a distorted circle shape) so we should not worry about distortion here.

**Comment by pllim** [comment]
For future me: With this, I don't need GWCS.

#### To downsample by 2

```python
w2_hdr['CRPIX1'] = 250 / 2
w2_hdr['CRPIX2'] = 150 / 2
w2_hdr['NAXIS1'] = 250
w2_hdr['NAXIS2'] = 150
w2_hdr['CDELT1'] = 2
w2_hdr['CDELT2'] = 2
```

#### To rotate

https://github.com/musevlt/mpdaf/blob/321cec4e95dc70c3e7eaf7402a2e3a24ebfb5b8e/lib/mpdaf/obj/coords.py#L1550

```python
def rotate(wcs_in, theta):
    wcs = deepcopy(wcs_in)
    theta = np.deg2rad(theta)
    sinq = np.sin(theta)
    cosq = np.cos(theta)
    mrot = np.array([[cosq, -sinq],
                     [sinq, cosq]])

    if wcs.wcs.has_cd():    # CD matrix
        newcd = np.dot(mrot, wcs.wcs.cd)
        wcs.wcs.cd = newcd
        wcs.wcs.set()
    elif wcs.wcs.has_pc():      # PC matrix + CDELT
        newpc = np.dot(mrot, wcs.wcs.get_pc())
        wcs.wcs.pc = newpc
        wcs.wcs.set()
    else:
        raise TypeError("Unsupported wcs type (need CD or PC matrix)")
    
    return wcs
```

**Comment by pllim** [comment]
The annulus background live-update is giving me a headache. To properly fix it, I have to convert sky aperture back to pixel aperture but in the frame of the selected data. So many events can trigger this update, and it has to update in the background (at least with how the code is currently set up) even when the background mode is something else.

Screenshot 2023-04-24 150830

Would be much easier (and more correct) to add annulus region support outside of this plugin first. 😩

File: notebooks/concepts/imviz_advanced_aper_phot.ipynb
Comment (pllim) [inline notebooks/concepts/imviz_advanced_aper_phot.ipynb:305]: @larrybradley , I don't understand why rectangle is not behaving. Might have to pick your brain when your power comes back on...

Screenshot 2023-08-08 175431

Comment (pllim) [reply notebooks/concepts/imviz_advanced_aper_phot.ipynb:305]: Figured it out. My bad https://github.com/spacetelescope/jdaviz/issues/2357

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (bmorris3) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:392]:
```suggestion
                f"Failed to calculate centroid: {repr(err)}", color='error', sender=self))
```
Not important, but `!r` confused me. I [read up on it](https://peps.python.org/pep-0498/#s-r-and-a-are-redundant) to confirm that it's equivalent to the above suggestion. Take it or leave it.

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue
Comment (bmorris3) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:85]: Per discussion in tag up/code review:
```suggestion
          :readonly="true"
          hint="Subset was defined with respect to this reference data (read-only)"
```

File: jdaviz/configs/imviz/helper.py
Comment (bmorris3) [inline jdaviz/configs/imviz/helper.py:222]:
```suggestion
        labels should be the reference data or look-up will fail.
```

File: jdaviz/configs/imviz/plugins/viewers.py
Comment (bmorris3) [inline jdaviz/configs/imviz/plugins/viewers.py:288]: If I understand your comment correctly, you're thinking about (near-)future-proofing this by looking up the current reference data's label? If so, I'll suggest that change here.

Comment (bmorris3) [inline jdaviz/configs/imviz/plugins/viewers.py:288]: This should work on this branch _and_ in https://github.com/spacetelescope/jdaviz/pull/2179:

```suggestion
        ref_label = self.state.reference_data.label
```

File: jdaviz/core/region_translators.py
Comment (bmorris3) [inline jdaviz/core/region_translators.py:54]:
```suggestion
    # the current viewer reference data, which can be changed.

```

File: jdaviz/core/region_translators.py
Comment (bmorris3) [inline jdaviz/core/region_translators.py:62]: Could be just: `to_sky = link_type == 'wcs'`

**Review by bmorris3 — COMMENTED** [review]
Neat!

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
Comment (pllim) [reply jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:392]: I guess this change is technically unrelated. I can revert it if it is too confusing.

File: jdaviz/configs/imviz/plugins/viewers.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/viewers.py:288]: But this reverts change from @kecnry in #1908

Comment (kecnry) [reply jdaviz/configs/imviz/plugins/viewers.py:288]: This is going to require some thought.... @bmorris3 in #2179 are links always updated wrt reference data?  I think until that is merged, we might need to leave this at `data_collection[0]`, but would love to be convinced otherwise.

Comment (bmorris3) [reply jdaviz/configs/imviz/plugins/viewers.py:288]: > in https://github.com/spacetelescope/jdaviz/pull/2179 are links always updated wrt reference data?

Links should be updated only when necessary in #2179.

I'm ok with holding off on the change and cleaning up if/when it's necessary in 2179. Feel free to mark as resolved.

**Review by bmorris3 — APPROVED** [review]
Thanks for the updates 👌🏻

File: jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue
Comment (kecnry) [inline jdaviz/configs/default/plugins/subset_plugin/subset_plugin.vue:87]: I think we'll eventually want to move this out of the subregions once reparenting is implemented (or perhaps remove entirely), but since its read-only, it doesn't hurt to have for now and could be useful for debugging.

File: jdaviz/configs/imviz/helper.py
Comment (kecnry) [inline jdaviz/configs/imviz/helper.py:222]: @bmorris3 - does the rotation work remove some of this flexibility (by forbidding mixed linking)?  If we might need to change or remove this, should we make it private in the meantime?

Comment (bmorris3) [reply jdaviz/configs/imviz/helper.py:222]: I think this method is useful with or without the rotation work. It's likely the case that it will be redundant with the link type attribute after image rotation is merged and WCS-linked in Imviz, but I think it's handy for debugging.

Comment (kecnry) [reply jdaviz/configs/imviz/helper.py:222]: Do you think it should be public or private in the meantime?

Comment (pllim) [reply jdaviz/configs/imviz/helper.py:222]: I don't see why it should be private. The information is so useful in the meantime that I think it should be public. If it comes a day we don't need it anymore, then we can deprecate it.

**Review by kecnry — APPROVED** [review]

**Comment by pllim** [comment]
There is a conflict now, I will have to rebase and let the CI run again.

**Comment by pllim** [comment]
Looks like a new deprecation warning from scipy over the weekend. It is unrelated.

```
`numpy.distutils` is deprecated since NumPy 1.23.0, as a result
```

**Pull request merged**
